trlx/trainer/nemo/gpt.py

```diff
@@ -0,0 +1,878 @@
+            batch_for_pipeline = None
+
+        # Pipeline stages will transfer this shape tensor to and from the
+        # previous and next stages
+        # The model must output a tensor of this shape if not the last pipeline
+        # stage. The model is given input of this shape if not the first pipeline
+        # stage via .set_input_tensor
+        tensor_shape = [
+            self.cfg.encoder_seq_length,
+            self.cfg.micro_batch_size,
+            self.cfg.hidden_size,
+        ]
+
+        # handle asynchronous grad reduction
+        if self.with_distributed_adam:
+            if self.megatron_amp_o2:
+                # copy grads to main grad
+                custom_sync_context_handler = lambda: self._optimizer.no_sync(
+                    greedy_grad_copy=True
+                )
+            else:
+                # keep grad tensors around
+                custom_sync_context_handler = lambda: self._optimizer.no_sync(
+                    greedy_grad_copy=False
+                )
+        else:
+            if self.megatron_amp_o2 and not self.cfg.get("sequence_parallel", False):
+                custom_sync_context_handler = self._optimizer.no_sync
+            else:
+                # TODO: enable async grad all reduce for O1/autocast mixed precision training
+                custom_sync_context_handler = None
+
+        # run forward and backwards passes for an entire global batch
+        # we do this inside training_step to support pipeline parallelism
+        # This gets the correct fwd/bwd pipeline step depending on the pipeline
+        # parallelism configuration
+        fwd_bwd_function = self._get_fwd_bwd_function()
+
+        losses_reduced_per_micro_batch = fwd_bwd_function(
+            forward_step_func=self.get_forward_output_and_loss_func(),
+            batch=batch_for_pipeline,
+            model=self.model,
+            forward_only=False,
+            tensor_shape=tensor_shape,
+            dtype=self.autocast_dtype,
+            grad_scaler=self.trainer.precision_plugin.scaler
+            if self.cfg.precision == 16
+            else None,
+            custom_sync_context_handler=custom_sync_context_handler,
+            sequence_parallel_enabled=self.cfg.get("sequence_parallel", False),
+            sync_batch_comm=self.cfg.get("sync_batch_comm", False),
+            num_micro_batches_with_partial_activation_checkpoints=self.cfg.get(
+                "num_micro_batches_with_partial_activation_checkpoints", None
+            ),
+        )
+
+        # only the last stages of the pipeline return losses
+        if losses_reduced_per_micro_batch:
+            # average loss across micro batches
+            loss_tensors_list = [
+                loss_reduced["avg"] for loss_reduced in losses_reduced_per_micro_batch
+            ]
+            loss_tensor = torch.concat(loss_tensors_list)
+            loss_mean = loss_tensor.mean()
```

Comment:
Can we collect the loss stats here e.g. `loss_q`, `loss_awac`, etc. for logging?